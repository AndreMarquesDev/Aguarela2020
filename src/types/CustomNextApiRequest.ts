import type { NextApiRequest } from 'next';
import type { FormPostRequestBody } from './FormPostRequestBody';

export type CustomNextApiRequest = {
    body: FormPostRequestBody;
} & NextApiRequest;
