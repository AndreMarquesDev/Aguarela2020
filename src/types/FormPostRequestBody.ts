import type { FormValues } from './FormValues';

export type FormPostRequestBody = {
    isPlaywright: boolean;
} & FormValues;
