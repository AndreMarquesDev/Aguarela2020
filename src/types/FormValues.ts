export type FormValues = {
    name: string;
    email: string;
    message: string;
    brand: string;
    subject: string;
};
