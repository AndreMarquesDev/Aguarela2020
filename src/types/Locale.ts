export enum Locale {
    Pt = 'pt',
    En = 'en',
}
