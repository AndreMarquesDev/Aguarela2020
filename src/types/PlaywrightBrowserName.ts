export enum PlaywrightBrowserName {
    Chromium = 'chromium',
    Firefox = 'firefox',
    Webkit = 'webkit',
}
