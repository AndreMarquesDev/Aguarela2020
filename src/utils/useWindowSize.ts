import { useEffect, useState } from 'react';

type WindowSize = {
    width: number;
    height: number;
};

export const useWindowSize = (): WindowSize => {
    const [windowSize, setWindowSize] = useState<WindowSize>({
        width: 0,
        height: 0,
    });

    useEffect(() => {
        const handleResize = (): void => {
            // eslint-disable-next-line react-hooks-extra/no-direct-set-state-in-use-effect
            setWindowSize({
                width: window.innerWidth,
                height: window.innerHeight,
            });
        };

        window.addEventListener('resize', handleResize);

        handleResize();

        return (): void => window.removeEventListener('resize', handleResize);
    }, []);

    return windowSize;
};

export enum Breakpoint {
    Desktop = 1280,
    Phablet = 1023,
    Tablet = 767,
    Mobile = 511,
}
