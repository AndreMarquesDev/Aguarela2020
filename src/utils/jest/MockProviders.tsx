/* eslint-disable react/jsx-no-constructed-context-values */
import React, { ReactNode } from 'react';
import { NavLinksContext } from '../../components/context/NavLinksContext';
import { TextsContext } from '../../components/context/TextsContext';
import { Locale } from '../../types/Locale';
import { textsEn, textsPt } from '../texts';

interface MockProvidersProps {
    language?: Locale;
    isMenuOpen?: boolean;
    toggleMenu?: jest.Mock;
    setNavHeight?: jest.Mock;
    children: ReactNode;
}

export const MockProviders = ({
    language = Locale.Pt,
    isMenuOpen = false,
    toggleMenu = jest.fn(),
    setNavHeight = undefined,
    children,
}: MockProvidersProps): JSX.Element => {
    return (
        <TextsContext.Provider
            value={{
                texts: language === Locale.Pt ? textsPt : textsEn,
            }}
        >
            <NavLinksContext.Provider value={{ isMenuOpen, toggleMenu, setNavHeight }}>
                {children}
            </NavLinksContext.Provider>
        </TextsContext.Provider>
    );
};
