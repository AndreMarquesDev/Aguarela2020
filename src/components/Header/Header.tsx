import type { Locale } from '../../types/Locale';
import type { Page } from '../../utils/pages';
import Image from 'next/image';
import Link from 'next/link';
import React, { useContext, useEffect, useRef } from 'react';
import { tabletMediaQuery } from '../../styles/mediaQueries';
import { headerDataTestId, homepageLogoLinkDataTestId } from '../../utils/dataTestIds';
import { Breakpoint, useWindowSize } from '../../utils/useWindowSize';
import { NavLinksContext } from '../context/NavLinksContext';
import { MenuIcon } from '../MenuIcon/MenuIcon';
import { NavLinks } from '../NavLinks/NavLinks';

export type HeaderProps = {
    currentRoute: string;
    language: Locale;
};

export const Header = ({ currentRoute, language }: HeaderProps): JSX.Element => {
    const { setNavHeight } = useContext(NavLinksContext);

    const navRef = useRef<HTMLDivElement>(null);

    useEffect(() => {
        if (setNavHeight && navRef.current) {
            setNavHeight(navRef.current.clientHeight);
        }
    });

    const { isMenuOpen, toggleMenu } = useContext(NavLinksContext);
    const handleMenuClick = (): void => toggleMenu(!isMenuOpen);

    const windowSize = useWindowSize();
    const isDesktop = windowSize.width > Breakpoint.Tablet;
    const isPhablet = windowSize.width < Breakpoint.Phablet;

    const logoWidth = isPhablet ? 150 : 250;

    return (
        <header data-testid={headerDataTestId}>
            <nav ref={navRef} className="wrapper">
                <Link
                    data-testid={homepageLogoLinkDataTestId}
                    href={`/${language}/${'homepage' as Page}`}
                    prefetch={false}
                >
                    <Image alt="Logo" height={35} src="/images/logo.svg" width={logoWidth} />
                </Link>

                <NavLinks
                    currentRoute={currentRoute}
                    isMenuOpen={isMenuOpen}
                    isMobile={!isDesktop}
                    language={language}
                />
                <MenuIcon isOpen={isMenuOpen} isVisible={!isDesktop} onClick={handleMenuClick} />
            </nav>

            <style jsx>
                {`
                    nav {
                        display: flex;
                        align-items: center;
                        justify-content: space-between;
                        padding-top: 60rem;
                        padding-bottom: 60rem;

                        @media (${tabletMediaQuery}) {
                            padding-top: 30rem;
                            padding-bottom: 30rem;
                        }
                    }
                `}
            </style>
        </header>
    );
};
