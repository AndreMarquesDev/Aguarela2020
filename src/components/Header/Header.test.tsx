import type { RenderResult } from '@testing-library/react';
import type { HeaderProps } from './Header';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Locale } from '../../types/Locale';
import { MockProviders } from '../../utils/jest/MockProviders';
import { setJestWindowWidth } from '../../utils/jest/setJestWindowWidth';
import { Breakpoint } from '../../utils/useWindowSize';
import { Header } from './Header';
import '@testing-library/jest-dom';

const baseProps: HeaderProps = {
    currentRoute: 'services',
    language: Locale.Pt,
};

const mockSetNavHeight = jest.fn();
const mockToggleMenu = jest.fn();

const renderComponent = (setNavHeight: jest.Mock | undefined): RenderResult => {
    return render(
        <MockProviders setNavHeight={setNavHeight} toggleMenu={mockToggleMenu}>
            <Header {...baseProps} />
        </MockProviders>,
    );
};

describe('<Header />', () => {
    test('renders properly', () => {
        const { container } = renderComponent(mockSetNavHeight);

        expect(mockSetNavHeight).toHaveBeenCalledTimes(2);

        expect(container).toMatchSnapshot();
    });

    test('renders properly below phablet', () => {
        setJestWindowWidth(Breakpoint.Mobile);

        const { container } = renderComponent(mockSetNavHeight);

        expect(container).toMatchSnapshot();
    });

    test('does not call the "setNavHeight" function if it is undefined', () => {
        renderComponent(undefined);

        expect(mockSetNavHeight).toHaveBeenCalledTimes(0);
    });

    test('calls the "toggleMenu" function when clicking on <MenuIcon />', async () => {
        const user = userEvent.setup();

        renderComponent(mockSetNavHeight);

        const menuIcon = screen.getByRole('button');

        expect(menuIcon).toBeInTheDocument();

        expect(mockToggleMenu).toHaveBeenCalledTimes(0);

        await user.click(menuIcon);

        expect(mockToggleMenu).toHaveBeenCalledTimes(1);
    });
});
