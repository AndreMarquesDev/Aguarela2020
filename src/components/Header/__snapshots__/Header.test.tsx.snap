// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders properly 1`] = `
<div>
  <header
    class="jsx-2405765474"
    data-testid="header_section"
  >
    <nav
      class="jsx-2405765474 wrapper"
    >
      <a
        data-testid="header_homepageLogoLink"
        href="/pt/homepage"
      >
        <img
          alt="Logo"
          data-nimg="1"
          decoding="async"
          height="35"
          loading="lazy"
          src="/images/logo.svg"
          style="color: transparent;"
          width="250"
        />
      </a>
      <ul
        class="jsx-2513558763 "
      >
        <li
          class="jsx-2513558763"
        >
          <a
            class="fontS navLinksLink"
            href="/pt/about"
          >
            Sobre
          </a>
        </li>
        <li
          class="jsx-2513558763"
        >
          <a
            class="fontS navLinksLink"
            href="/pt/projects"
          >
            Projetos
          </a>
        </li>
        <li
          class="jsx-2513558763"
        >
          <a
            class="fontS active navLinksLink"
            href="/pt/services"
          >
            Serviços
          </a>
        </li>
        <li
          class="jsx-2513558763"
        >
          <a
            class="fontS navLinksLink"
            href="/pt/contact"
          >
            Contacto
          </a>
        </li>
        <li
          class="jsx-2513558763 languageButton fontXXS"
        >
          <button
            class="jsx-2866958827"
            type="button"
          >
            pt
          </button>
        </li>
      </ul>
      <i
        aria-label="toggle menu"
        class="jsx-2279021353 menuIcon hidden"
        data-testid="menuIcon_button"
        role="button"
        tabindex="0"
      >
        <span
          class="jsx-2279021353"
        />
      </i>
    </nav>
  </header>
</div>
`;

exports[`<Header /> renders properly below phablet 1`] = `
<div>
  <header
    class="jsx-2405765474"
    data-testid="header_section"
  >
    <nav
      class="jsx-2405765474 wrapper"
    >
      <a
        data-testid="header_homepageLogoLink"
        href="/pt/homepage"
      >
        <img
          alt="Logo"
          data-nimg="1"
          decoding="async"
          height="35"
          loading="lazy"
          src="/images/logo.svg"
          style="color: transparent;"
          width="150"
        />
      </a>
      <ul
        class="jsx-2513558763 mobileLayout"
      >
        <li
          class="jsx-2513558763"
        >
          <a
            class="fontS navLinksLink"
            href="/pt/about"
          >
            Sobre
          </a>
        </li>
        <li
          class="jsx-2513558763"
        >
          <a
            class="fontS navLinksLink"
            href="/pt/projects"
          >
            Projetos
          </a>
        </li>
        <li
          class="jsx-2513558763"
        >
          <a
            class="fontS active navLinksLink"
            href="/pt/services"
          >
            Serviços
          </a>
        </li>
        <li
          class="jsx-2513558763"
        >
          <a
            class="fontS navLinksLink"
            href="/pt/contact"
          >
            Contacto
          </a>
        </li>
        <li
          class="jsx-2513558763 languageButton fontXXS white fontXS"
        >
          <button
            class="jsx-2866958827"
            type="button"
          >
            pt
          </button>
        </li>
      </ul>
      <i
        aria-label="toggle menu"
        class="jsx-2279021353 menuIcon"
        data-testid="menuIcon_button"
        role="button"
        tabindex="0"
      >
        <span
          class="jsx-2279021353"
        />
      </i>
    </nav>
  </header>
</div>
`;
